### DISCUSSION TITLE
Query not working: Index boost with map reduce

### DISCUSSION THREAD

**Discussion by Yuzal** [ravendb/ravendb#12910] [Q&A] [answered]
Current Situation: 
I have a multi map index with a reduce result.
while querying it i am sorting these results by lucene score (OrderByScore).

New Situation:
Now i want the lucene score to ignore a specific property and still be able to query or store the specific property.

i am not able to query for the property properly when i use the index boost.
Quering in rql for the property seems to return me boostedValue objects.

How do i handle this situation properly?

**Comment by krivulcik** [comment, marked as answer]
Something like this works nicely in our case:

```
from index 'YourIndex'
where search(Name, "test")
and boost(search(ScoreIgnoredField, "test2"), 0)
order by score()
include explanations()
```

Sample score explanation:

```
16.386274 = (MATCH) sum of:
  16.386274 = (MATCH) fieldWeight(Name:test in 76582), product of:
    1 = tf(termFreq(Name:test)=1)
    16.386274 = idf(docFreq=15, maxDocs=7965161)
    1 = fieldNorm(field=Name, doc=76582)
  0 = (MATCH) weight(ScoreIgnoredField:test2^.0 in 76582), product of:
    0 = queryWeight(ScoreIgnoredField:test2^.0), product of:
      0 = boost
      4.6887517 = idf(docFreq=124374, maxDocs=7965161)
      0.061026685 = queryNorm
    4.6887517 = (MATCH) fieldWeight(ScoreIgnoredField:test2 in 76582), product of:
      1 = tf(termFreq(ScoreIgnoredField:test2)=1)
      4.6887517 = idf(docFreq=124374, maxDocs=7965161)
      1 = fieldNorm(field=ScoreIgnoredField, doc=76582)
```

Only the `Name` field score is taken into account.

**Reply by Yuzal** [reply]
unfortunately it doesnt cover my situation.
how would you handle this situation:

from index "dummyIndex"
where (boost(DepartmentsList in("deparmentId1", "departmentId2", "deparmentId3")), 0)
order by score(), creationDate desc

Currently i will always get documents first with the highest departments counts and never the newest ones

**Reply by krivulcik** [reply]
Are you filtering by anything else (not in this sample query)?

If not, this should be enough:

```
from index "dummyIndex"
where DepartmentsList in("deparmentId1", "departmentId2", "deparmentId3")
order by creationDate desc
```

The boosting and ordering by the score are not needed in such a case.

**Reply by Yuzal** [reply]
it could have several additional filter definied by a user therefore i am using the lucene scoring

**Reply by ayende** [reply]
The `OrderByScore()` is only relevant if you are trying to find things that the user is interested in based on the query.
That tends to be pretty rare, to be honest. In almost all cases, you want to sort by something that the user needs.

For example, `order by CreationDate desc` is the most common one.
Doing a query by both `score()` and `CreationDate` is really strange, because what you are saying is that you want to first sort by the `score()` and if there are documents with the same score, use the `CreationDate` as the tie breaker.

Can you explain more about the scenario?

**Reply by Yuzal** [reply]
I have several optional filter and a full text search.
the user itself has the option to sort by a defined parameter (creationDate etc) or by scoring.
the scoring is usually only important when he uses the full text search

**Reply by ayende** [reply]
Given that scenario, and the ability to set boost to `0`, what do you need to be able to build your feature?

**Reply by krivulcik** [reply]
I believe that @Yuzal 's issue can be demonstrated with the following failing test.

It seems that the boosting "in" to 0 affects something that shouldn't be affected.

```
using Raven.Client.Documents;
using Raven.Client.Documents.Indexes;
using Raven.Client.Documents.Linq;
using Raven.TestDriver;
using System.Linq;
using Xunit;

namespace RavenDB5Tests.Tests
{
    public class ZeroBoostOrdering : RavenTestDriver
    {
        [Fact]
        public void ZeroBoostOrderingTest()
        {
            using var store = GetDocumentStore();

            new DocIndex().Execute(store);

            using (var session = store.OpenSession())
            {
                session.Store(new Doc { Id = "doc-1", StrVal = "match", NumVal = 2 });
                session.Store(new Doc { Id = "doc-2", StrVal = "better match", NumVal = 1 });

                session.SaveChanges();
            }

            WaitForIndexing(store);
            WaitForUserToContinueTheTest(store);

            using (var session = store.OpenSession())
            {
                var query = session.Advanced.DocumentQuery<Doc, DocIndex>()
                    .WhereIn(x => x.StrVal, new[] { "better", "match" }).Boost(0)
                    .ToQueryable()
                    .OrderByScore()
                    .ThenByDescending(x => x.NumVal);

                // Another way to create the query:
                //var query = session.Advanced.RawQuery<Doc>(
                //    @"from index 'DocIndex'
                //    where boost(StrVal in ('better', 'match'), 0)
                //    order by score(), NumVal desc");

                var results = query.ToArray();

                Assert.Equal(2, results.Length);
                Assert.Equal("doc-1", results[0].Id);
                Assert.Equal("doc-2", results[1].Id);
            }
        }

        public class DocIndex : AbstractIndexCreationTask<Doc>
        {
            public DocIndex()
            {
                Map = docs =>
                    from doc in docs
                    select new
                    {
                        doc.Id,
                        doc.StrVal,
                        doc.NumVal
                    };

                Indexes.Add(x => x.StrVal, FieldIndexing.Search);
            }
        }

        public class Doc
        {
            public string Id { get; set; }
            public string StrVal { get; set; }
            public int NumVal { get; set; }
        }
    }
}
```

**Reply by ayende** [reply]
`in` query isn't a full text search one, it doesn't take into account boosting, since it is a constant match.
If you do the same with a `search()`, it will properly work, on the other hand.

**Reply by Yuzal** [reply]
> `in` query isn't a full text search one, it doesn't take into account boosting, since it is a constant match. If you do the same with a `search()`, it will properly work, on the other hand.

okay so wherein doesnt accept boosting while a normal where is working with a boost?
i just checked our explanations and it seems to ignore wherein boosting but working with everything else (full text, where etc)

**Reply by krivulcik** [reply]
That makes sense. Should the ordering in this case work as it does? When the score is the same in both records, I would expect the second ordering criterion to break the tie.

I'm not sure whether using `search()` is appropriate for @Yuzal 's use case, he seems to search in an array or something similar.

**Reply by ayende** [reply]
It looks like it is working fine. But note that the scenario where you are doing searches like that is pretty strange. There is intentionally little control over the scoring, so usually it is the only item there.

**Reply by krivulcik** [reply]
Not sure, if it's apparent from the test, and it's maybe even moot, given that the original problem is now resolved.

But given a query with the following sorting: `order by score(), NumVal desc` , it is surprising to get the results in the following order:

```
doc-2 (score = 1; NumVal = 1)
doc-1 (score = 1; NumVal = 2)
```

The score is the same in both cases, but the NumVal is not ordered as requested in the query.

In this case, it might be strange, as I tried to isolate the issue, and it manifests itself only with `in`. However, when it would be part of a more complex query, it might cause unexpected outcomes overall, I think.

**Reply by Yuzal** [reply]
I am not entirely sure what do you mean with constant match?

from index 'MyIndex'
where boost(Collection in("Entity/4ffdc8b6-ac63-4143-96c5-05ce8f77bf44"), 0)
=> No boost working

from index 'MyIndex'
where boost(State == true, 0)
=> boost is working

To explain my situation:
We have a database with applicants (15.000+) and a user has only limited access to applicants.
This limitation is regulated by his assigned branches.
He can search the database with full search and has several other filter options as well.

Currently everytime he is doing a search by scoring, the number of branches of the applicants are scored as well despite boosting to 0:

Edit:
from index 'MyIndex'
where boost(Collection in("Entity/4ffdc8b6-ac63-4143-96c5-05ce8f77bf44"), 0)
and boost(State == true, 0)
and search(Searchfield, "userinput")

**Reply by ayende** [reply]
@krivulcik About `in` query & boosting, this is fixed here: https://github.com/ravendb/ravendb/pull/12949

@Yuzal I think that the PR should resolve this for you.
